Add language selector to change the app language

B0pol opened the pull request:
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

Fix #2794, fix #2162, fix #2887, fix #1215, fix #1130, fix [r/NewPipe : Any way to change the app language?](https://www.reddit.com/r/NewPipe/comments/b8sqpf/any_way_to_change_the_app_language/), [r/NewPipe : how to change app language ?](https://www.reddit.com/r/NewPipe/comments/elamrg/how_to_change_app_language/)

Explaination :
Before, NewPipe's language was phone's language. Now, it's the language selected in Settings -> Content -> App language, and by default, it stays system default language.
Thus, you can force an app language (eg. English for developing / debugging : take a screenshot in English) or use non-available system language (Esperanto isn't translated for Android) or simply just set the language you want **without changing phone / system language**.

APK: [2020-01-27_language-selector#2921_APK.zip](https://github.com/TeamNewPipe/NewPipe/files/4118734/2020-01-27_language-selector.2921_APK.zip)

B0pol commented:
I found out what does content language and also I found it's not setting the language you want sometimes :
If the language is like language-COUNTRY, for example en-GB (UK english), it doesn't work and use phone's language. Same if the language in the list is not translated : actually, the list is used for search and other stuff. 
Then think we should use a third selector, see this : 
![Screenshot_1578325520](https://user-images.githubusercontent.com/58657617/71829034-9c62b480-309b-11ea-9cf0-8c9b84a73d28.png)
where we would set the language already translated in NewPipe, + at the top a button that would be : system language.

I'll work on it this week

B0pol commented:
I finished the work and tested all languages. I found an exception (#2940), but it's not related to my PR, it's related to the formatter and %1$s not in the good order and arabic alphabet, then you can merge this PR.

B0pol commented:
I changed the way time is displayed, it now use NewPipe's language (the new selector).
This fixes #1215 . In the limit of the languages of the library Pretty time we use : https://www.ocpsoft.org/prettytime/#section-12
I'll try to use custom TimeFormat with Esperanto (the library supports to add languages), and if it works, it will be translatable with weblate for all the other languages.

B0pol commented:
The PR is ready for a review, please consider doing it @Stypox @TobiGr @mauriciocolli
BTW, here is some screenshots :
System language was Spanish
[US trendings displayed in English](https://user-images.githubusercontent.com/58657617/72161936-ebb92580-33b8-11ea-927c-9f88d26f4f5b.png) | [Settings in English](https://user-images.githubusercontent.com/58657617/72162178-5bc7ab80-33b9-11ea-98e9-30143aa4afe8.png)

[US trendings displayed in German](https://user-images.githubusercontent.com/58657617/72161986-07243080-33b9-11ea-9640-0e13a31e6eaf.png) | [Settings in German](https://user-images.githubusercontent.com/58657617/72162018-15724c80-33b9-11ea-828e-e1145627a244.png)

[US trendings displayed in System language, Spanish](https://user-images.githubusercontent.com/58657617/72162273-92052b00-33b9-11ea-966d-e364c44331f4.png) | [Settings in System language, Spanish](https://user-images.githubusercontent.com/58657617/72162134-4b173580-33b9-11ea-9132-938985540e18.png)

B0pol commented:
I just fixed Indonesian that wasn't possible to select by renaming values-id to values-in.
Briefly, android use outdated version of an iso norm, then it doesn't recognize id, but in.
See https://stackoverflow.com/questions/13291578/how-to-localize-an-android-app-in-indonesian-language.
I also added Occitan to the selector.
So the selector is working for every language now.
Pirate will show English as [values-pr/strings.xml](https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/res/values-pr/strings.xml) is empty, and for some languages you will see a lot of english, as it's not translated much (I'm thinking to Interlingua, Occitan…)

B0pol commented:
APK: [app-debug-language-selector_2020-01-18.zip](https://github.com/TeamNewPipe/NewPipe/files/4080229/app-debug-language-selector_2020-01-18.zip)

B0pol commented:
@abusarimhindi APK: [apk-debug](https://github.com/TeamNewPipe/NewPipe/files/4108063/app-debug.zip)
In settings, you can change for Urdu in settings / content settings / NewPipe's language.

For anyone else wondering, this is a TEST version: there might be some problems, because I was based on dev branch in extractor, and I switched to master branch of extractor (the only one that includes the fix for decryption bug)

I manually updated extractor version so that YouTube works again, and remove the urdu string, took the one from weblate.

abusarimhindi commented:
> APK: [app-debug-language-selector_2020-01-19.zip](https://github.com/TeamNewPipe/NewPipe/files/4082971/app-debug-language-selector_2020-01-19.zip)
Thanks!
But، I think this is not updated version (0.18.2), and Urdu updated strings are not putted in this file.

![Annotation 2020-01-24 175654](https://user-images.githubusercontent.com/22712698/73069227-32d40a00-3ed3-11ea-80a0-8369a4a3d167.png)

B0pol commented:
This is 0.18.1 because I'm based on debug build, which is still on 0.18.1.
But as I said, I manually updated extractor version, to prevent the decryption crash, and I manually updated urdu strings from weblate, just for you, and I didn't commit these changes because this should be fixed by updating dev branch, not by fixing manually

abusarimhindi commented:
> This is 0.18.1 because I'm based on debug build, which is still on 0.18.1.
> But as I said, I manually updated extractor version, to prevent the decryption crash, and I manually updated urdu strings from weblate, just for you, and I didn't commit these changes because this should be fixed by updating dev branch, not by fixing manually

Sorry, It is Ok. Firstly I downloaded the old file, now updated. Thanks again.

B0pol commented on app/src/main/java/org/schabi/newpipe/util/Localization.java:
```diff
@@ -199,21 +204,41 @@ public static String getDurationString(long duration) {
+        Locale loc;
+        if (lang.equals("system")) {
+            loc = Locale.getDefault();
+        } else if (lang.matches(".*-.*")) {
```
not en-gb and en-us (there is only one form of english in NewPipe and weblate), but pt (portugal) from pt-br (brazil), same for nl and nl-be, chinese… I'll do this in the next commit

B0pol commented:
> ```java
> public static void assureCorrectAppLanguage() {
>     changeAppLanguage(getAppLocale(getApplicationContext()), getResources());
> }
> ```

I can't do that, I can't have context and resources from a static context.
The most optimized thing I can have is
```java
public static void assureCorrectAppLanguage(Context c) {
    changeAppLanguage(getAppLocale(c), c.getResources())
}
```
and then, in mainactivity and other activites needing it:
``assureCorrectAppLanguage(this);`` or ``assureCorrectAppLanguage(getContext());`` sometimes.
As assureCorrectAppLanguage(getApplicationContext()); doesn't work, ill need time for testing.

> Is it necessary to call changeAppLanguage() in all onCreate and onResume methods or can we just do it in App.java in?

Yesn't sir! If you put in onCreate and onResume of mainActivity, it will work for about 95%. I'll say why I put for the other files:
AboutActivity: everything is fine because strings are loaded on runtime, except title which is set in manifest.xml, so I changed it on runtime

Player: kinda tough, it is in the language you want, except if you press a button, change orientation (landscape / portrait).

The other file honestly I don't remember, but at first I put in MainActivity, and while testing phase I added where it would show the wrong language.
It's kinda like the ``ThemeHelper.setTheme(this, ServiceHelper.getSelectedServiceId(this));``: you don't need _everywhere_, but in activities who are making problems without it.

I've tested for a long time since I opened the PR, then I'm almost sure there won't be those issues with the current build, but if someone found one later, you just need to add ``assureCorrectAppLanguage(getApplicationContext());``, and as I said, it works most of the time, as I said by putting only in MainActivity it works almost everywhere.

TobiGr commented:
> > ```java
> > public static void assureCorrectAppLanguage() {
> >     changeAppLanguage(getAppLocale(getApplicationContext()), getResources());
> > }
> > ```
> 
> I can't do that, I can't have context and resources from a static context.
> The most optimized thing I can have is
> 
> ```java
> public static void assureCorrectAppLanguage(Context c) {
>     changeAppLanguage(getAppLocale(c), c.getResources());
> }
> ```

My bad. This happens when you write reviews during lectures :D

TobiGr commented:
Code looks go so far. I'll test it as soon as possible

B0pol commented:
Just one thing you should know
I made changes to values-ur/strings.xml, because of formatter crash (#2940)
They were fixed by weblate, so either you update weblate soon and I solve conficts before you merge the pr, or for every conflicts existing with Urdu strings, you should pick the Weblate version

TobiGr commented:
> I made changes to values-ur/strings.xml, because of formatter crash (#2940)
They were fixed by weblate in the meantime, so either you update dev with weblate soon and I solve conficts before you merge the pr, or for every future conflicts existing with Urdu strings, you should pick the Weblate version.

I merged weblate into dev. Could you take a look at the conflicts? I am short on time until I finished my exams. Thanks :)

B0pol commented:
> I merged weblate into dev. Could you take a look at the conflicts? I am short on time until I finished my exams. Thanks :)

Done. Actually, the only conflict shown on Github was the indonesian file which have been renamed, but while merging with git I had no problem.

I think the PR is finally fully ready.

TobiGr commented:
Thanks. One small thing. Can you please integrate this setting into the mechanism which displays a toast that language changes require an app restart?

https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L135-L140

B0pol commented:
> Thanks. One small thing. Can you please integrate this setting into the mechanism which displays a toast that language changes require an app restart?
> 
> https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L135-L140

It's already done. Look at https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L135-L136 and https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L71

B0pol commented:
> Thanks. One small thing. Can you please integrate this setting into the mechanism which displays a toast that language changes require an app restart?
> 
> https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L135-L140

done
